### PR TITLE
Don't capture BoundNode when creating a LazyNoneOperation

### LINKED ISSUE #34918: CSharpOperationFactory `CreateInternal` Should Not Capture
https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs#L321

This captures the `boundNode` element in the lambda for `getChildren`, and it doesn't need to. We should update this to be non-capturing.

I'm going to mark this up for grabs so that anyone in the community who would like to make a small improvement can do so, as I am focusing on the nullable api at the moment. If no has taken this up after that work is completed, I'll probably fix this.

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Fixes #34918

### CODE REVIEW

**Comment by YairHalberstadt** [comment]
@333fred

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs [hunk 1]
```diff
@@ -10,6 +11,23 @@
 
 namespace Microsoft.CodeAnalysis.Operations
 {
+
+    internal sealed class CSharpLazyNoneOperation : LazyNoneOperation
+    {
+        private readonly Func<BoundNode, ImmutableArray<IOperation>> _getChildren;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs]: This is still capturing. We should instead hold onto the `CSharpOperationFactory` here, and call `GetIOperationChildren` off that instance below.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationNodes.vb [hunk 2]
```diff
@@ -5,6 +5,23 @@ Imports System.Threading
 Imports Microsoft.CodeAnalysis.VisualBasic
 
 Namespace Microsoft.CodeAnalysis.Operations
+    Friend NotInheritable Class VisualBasicLazyNoneOperation
+        Inherits LazyNoneOperation
+
+        Private ReadOnly _getChildren As Func(Of BoundNode, ImmutableArray(Of IOperation))
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationNodes.vb]: Same comment as C#.

**Comment by YairHalberstadt** [comment]
@333Fred
The CSharpOperationFactory calls this with two different methods. It doesn't always use GetIOperationChildreb.

**Comment by 333fred** [comment]
Then what we can do is have `GetIOperationChildren` check to see if the node passed in is a `BoundMultipleLocalDeclarations`, and have it do the correct thing in that case.

**Comment by YairHalberstadt** [comment]
Done.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 3]
```diff
@@ -2026,13 +2034,17 @@ private IOperation CreateBoundDiscardPatternOperation(BoundDiscardPattern boundN
         private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarations boundNode)
         {
             //TODO: Implement UsingLocalDeclaration operations correctly.
-            //      For now we return an implicit operationNone, with a single child consisting of the using declaration parsed as if it were a standard variable declaration
+            //      For now we return an implicit operationNone, 
+            //      and GetIoperationChildren will return a single child
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >o [](start = 28, length = 1)

Nit: spelling

Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Also trailing whitespace.

---
In reply to: [289972989](https://github.com/dotnet/roslyn/pull/36055#discussion_r289972989) [](ancestors = 289972989)

Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: fixed, and fixed, assuming I got the right spelling mistake :-)

**Comment by 333fred** [comment]
Thanks @YairHalberstadt!

**Pull request merged**